rb3ptr/rb3ptr.3
.TH rb3ptr 3 "May 25, 2017"
.SH NAME
rb3ptr \- Generic Red-black tree implementation and macros for typesafe access
.SH SYNOPSIS

The rb3ptr API is split into multiple sections.

DATATYPES contains essential structures and definitions.
.LP
BASIC FUNCTIONS has operations on the generic Red-black tree implementation
that provide an ordered container API.
.LP
NAVIGATION FUNCTIONS contains additional functionality for navigation in a
binary search tree.
.LP
MACROS explains the type-specific wrappers for better type-safety and more
convenient access with fixed comparison functions.
.LP
BSD MACROS lists functionality to emulate parts of the BSD
.I <sys/tree.h>
API.


.SS DATATYPES

.BI "struct rb3_tree;"
.LP
.BI "struct rb3_head;"
.LP
.BI "typedef int (*rb3_cmp)(struct rb3_head *" head ", void *" data);


.SS BASIC FUNCTIONS

.BI "void rb3_init(struct rb3_tree *" tree );
.LP
.BI "void rb3_exit(struct rb3_tree *" tree );
.LP
.BI "int rb3_isempty(struct rb3_tree *" tree );
.LP
.BI "struct rb3_head *rb3_get_min(struct rb3_tree *" tree );
.LP
.BI "struct rb3_head *rb3_get_max(struct rb3_tree *" tree );
.LP
.BI "struct rb3_head *rb3_get_prev(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_get_next(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_get_minmax(struct rb3_tree *" tree ", int " dir );
.LP
.BI "struct rb3_head *rb3_get_prevnext(struct rb3_head *" head ", int " dir );
.LP
.BI "struct rb3_head *rb3_find(struct rb3_tree *" tree ", struct rb3_cmp " cmp ", void *" data );
.LP
.BI "struct rb3_head *rb3_find_parent(struct rb3_tree *" tree ", struct rb3_cmp " cmp ", void *" data ", struct rb3_head **" parent_out ", int *" dir_out );
.LP
.BI "void rb3_link_and_rebalance(struct rb3_head *" head ", struct rb3_head *" parent ", int " dir );
.LP
.BI "void rb3_unlink_and_rebalance(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_insert(struct rb3_tree *" tree ", struct rb3_head *" head ", struct rb3_cmp " cmp ", void *" data);
.LP
.BI "struct rb3_head *rb3_delete(struct rb3_tree *" tree ", struct rb3_cmp " cmp ", void *" data );


.SS NAVIGATION FUNCTIONS

Note: Valid values for
.I dir
are
.B RB3_LEFT
and
.B RB3_RIGHT
(0 and 1).

.BI "struct rb3_head *rb3_get_root(struct rb3_tree *" tree );
.LP
.BI "struct rb3_head *rb3_get_base(struct rb3_tree *" tree );
.LP
.BI "int rb3_is_base(struct rb3_head *" head );
.LP
.BI "int rb3_is_node_linked(struct rb3_head *" head );
.LP
.BI "int rb3_get_parent_dir(struct rb3_head *" head );
.LP
.BI "int rb3_has_child(struct rb3_head *" head ", int " dir );
.LP
.BI "struct rb3_head *rb3_get_parent(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_get_child(struct rb3_head *" head ", int " dir );
.LP
.BI "struct rb3_head *rb3_get_prev_ancestor(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_get_next_ancestor(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_get_prev_descendant(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_get_next_descendant(struct rb3_head *" head );
.LP
.BI "struct rb3_head *rb3_get_prevnext_ancestor(struct rb3_head *" head ", int " dir );
.LP
.BI "struct rb3_head *rb3_get_prevnext_descendant(struct rb3_head *" head ", int " dir );


.SS MACROS

.BI RB3_GEN_IMPL()
.LP
.BI RB3_GEN_IMPL_STATIC()
.LP
.BI RB3_GEN_INLINE( NAME ", " NODETYPE ", " GET_HEAD ", " GET_NODE )
.LP
.BI RB3_GEN_NODECMP( NAME ", " NODETYPE ", " GET_HEAD ", " GET_NODE ", " NODECMP )
.LP
.BI RB3_FOREACH( NAME ", " TREE ", " NODE );
.LP
.BI RB3_FOREACH_REVERSE( NAME ", " TREE ", " NODE );
.LP
.BI RB3_FOREACH_DIR( NAME ", " TREE ", " NODE );
.LP
.BI RB3_FOREACH_SAFE( NAME ", " TREE ", " NODE ", " TMPNODE );
.LP
.BI RB3_FOREACH_REVERSE_SAFE( NAME ", " TREE ", " NODE ", " TMPNODE );
.LP
.BI RB3_FOREACH_DIR_SAFE( NAME ", " TREE ", " NODE ", " TMPNODE );


.SS BSD MACROS

These MACROS expose a BSD
.I <sys/tree.h>
compatible interface. Unfortunately,
.BR RB_LEFT ()
and
.BR RB_RIGHT ()
cannot be supported due to missing information in the signature.

.BI RB_PROTOTYPE()
.LP
.BI RB_PROTOTYPE_STATIC()
.LP
.BI RB_GENERATE()
.LP
.BI RB_GENERATE_STATIC()
.LP
.BI RB_INIT( tree )
.LP
.BI RB_INSERT( NAME ", " tree ", " elm )
.LP
.BI RB_FIND( NAME ", " tree ", " elm )
.LP
.BI RB_REMOVE( NAME ", " tree ", " elm )
.LP
.BI RB_MIN( NAME ", " tree )
.LP
.BI RB_MAX( NAME ", " tree )
.LP
.BI RB_PREV( NAME ", " tree ", " elm )
.LP
.BI RB_NEXT( NAME ", " tree ", " elm )


.SH DOCUMENTATION

This section contains explanations for the structures and prototypes listed
above.

.SS DATATYPES

.B "struct rb3_tree"
is the basic tree type. It holds the root link for one red-black tree in a
running program.

.B "struct rb3_head"
is the linking information for a node in the tree. Data that should be linked
in a tree must contain such a structure. The tree implementation does not care
about the actual data, but simply maintains the links between the link
structures.

.B rb3_cmp
is the function type of user-provided comparisons to direct tree searches. At
each visited node, the function is called with the node and a user-provided
data as arguments. It should return an integer less than, equal to, or greater
than 0, indicating whether the node in the tree compares less than, equal to,
or greater than the user-provided data. This function is always user-provided.
Typically it will make use
.BR offsetof "(3) or the linux " container_of ()
macro to get at the actual data in which the
.B struct rb3_head
node is embedded.


.SS BASIC FUNCTIONS

.BR rb3_init ()
initializes a
.B struct rb3_tree.
Zeroing out the structure (e.g., with
.BR memset()
or static initialization) is also believed to work.

.BR rb3_exit ()
frees the resources associated with a tree. There are currently no such
resources, and it should also be ok to zero it out. Usage of this function is
recommended for future-proofness and also for reasons of symmetry.

.BR rb3_isempty ()
tests if a tree does not contain any nodes. This of course is true after
initialization.

.BR rb3_get_min "() and " rb3_get_max ()
return the leftmost / rightmost element linked in a tree. If the tree is empty,
NULL is returned.

.BR rb3_get_prev "() and " rb3_get_next ()
return the previous / next node linked in the same tree (with respect to
in-order traversal). If no such node exists, NULL is returned.

.BR rb3_get_minmax "() and " rb3_get_prevnext ()
can be used instead of
.BR rb3_get_min "(), " rb3_get_max "(), " rb3_get_prev "(), and " rb3_next () .
They take the direction as runtime parameter
.RB ( RB3_LEFT " or " RB3_RIGHT ).


.BR rb3_find ()
finds a node in a tree. If no node comparing equal (i.e., the comparison
function returns 0 given the visited node and the user-provided data) is found
in the tree, NULL is returned.

.BR rb3_find_parent ()
is similar to
.BR rb3_find (),
but when the search is unsuccessful, the appropriate insertion point for a node
matching the search is returned in the out-arguments.
.BR rb3_link_and_rebalance ()
can then be used to add the node.
.RB ( rb3_insert ()
combines these two operations in a single function call).

.BR rb3_link_and_rebalance ()
can be used to link a given node into a tree given an insertion point (parent
node and its child direction). The appropriate insertion point can be found
using
.BR rb3_find_parent ().

.BR rb3_unlink_and_rebalance ()
can be used to unlink a given node from a tree without any search. The node
must be known to be linked in a tree.

.BR rb3_insert ()
can be used to insert a new node into a tree at a suitable insertion point. It
takes a tree, the new node to insert, and a
.B rb3_cmp
function implementing the node ordering to direct the search. If a node
comparing equal (i.e., the comparison function returns 0 given the visited node
and the user-provided node) is found in the tree, that node is returned.
Otherwise, the to-be-inserted node is linked into the tree and NULL is
returned.

.BR rb3_delete ()
does a node search in a tree given a comparison function and data. If a
matching node is found, it is unlinked from the tree and a pointer to it is
returned. Otherwise, NULL is returned.


.SS NAVIGATION FUNCTIONS

.BR rb3_get_root ()
returns the root node in the tree, or NULL if the tree is empty.

.BR rb3_get_base ()
returns the base head of the tree, which always exists. If the tree is
nonempty, the root node is linked as left child of the base node. This is an
implementation detail and need not be relied upon in most situations.

.BR rb3_is_base ()
tests whether a link structure is the base node in a tree. This only can distinguish the base node of a tree that was initialized with
.BR rb3_init (),
from non-base nodes that are zeroed or properly linked in a tree.

.BR rb3_is_node_linked ()
tests whether the given non-base node is linked in a (any) tree. This can only
distinguish nodes that are properly linked in a tree from unlinked or zeroed
nodes.

.BR rb3_get_parent_dir ()
returns RB3_LEFT or RB3_RIGHT depending on whether the given link node is the
left or right child of its parent. This is a single bitwise operation on the
link structure, so is more efficient than testing both childs of the parent's
link structure.

.BR rb3_has_child ()
tests whether the given link has a child in the given direction.

.BR rb3_get_parent ()
returns the parent link structure of the given node. If the given node is the
root node, the base head is returned. If this is not what you want, test if the
return value has itself a parent. (The base head is the only head that has no
parent).

.BR rb3_get_child ()
returns the left or right child of the given node, depending on the given
direction value
.RB ( RB3_LEFT " or " RB3_RIGHT )

.BR rb3_get_prev_ancestor ()
returns the nearest left ancestor of the given head link structure. If none
exists, NULL is returned.

.BR rb3_get_next_ancestor ()
returns the nearest right ancestor of the given head link structure. If none
exists, NULL is returned.

.BR rb3_get_prev_descendant ()
returns the nearest left descendant of the given head link structure. If none
exists, NULL is returned.

.BR rb3_get_next_descendant ()
returns the nearest right descendant of the given head link structure. If none
exists, NULL is returned.

.BR rb3_get_prevnext_ancestor ()
returns the nearest left or right ancestor (depending on the given direction)
of the given head link structure. If none exists, NULL is returned.

.BR rb3_get_prevnext_descendant ()
returns the nearest left or right descendant (depending on the given direction)
of the given head link structure. If none exists, NULL is returned.


.SS MACROS

.BR RB3_GEN_IMPL ()
evaluates to a complete implementation of the rb3ptr API with
.I extern
linkage. Use this only if you can't use a separately compiled rb3ptr libray.
Macros are hard to debug.

.BR RB3_GEN_IMPL_STATIC ()
evaluates to a complete implementation of the rb3ptr API with
.I static
linkage. Use this only if no other file in the same projects need rb3ptr's
functionality.

.BR RB3_GEN_INLINE ()
evaluates to an implementation of the non-comparison-related functionality of
rb3ptr wrapped for a specific datatype.
.I NAME
should be a prefix for these functions, such as for example
.IR footree .
.I NODETYPE
should be the node type managed by this set of generated functions, such as
for example
.IR struct foo
(see the example below).
.I GET_HEAD " and " GET_NODE
should be macros or functions for the generated implementation's use to
retrieve the embedded link structure from a node, or vice versa.

.BR RB3_GEN_NODECMP ()
TODO

.BR RB3_FOREACH ()
is a for-loop iteration macro.
.I NAME
should be the prefix used in
.IR RB3_GEN_INLINE ().
.I TREE
should be a tree of the generated type
.RB ( "struct NAME" ).
.I NODE
should by a value of type
.IR "NODETYPE *" .
It is used as iteration variable.


.BR RB3_FOREACH_REVERSE ()
.BR RB3_FOREACH_DIR ()
.BR RB3_FOREACH_SAFE ()
.BR RB3_FOREACH_REVERSE_SAFE ()
.BR RB3_FOREACH_DIR_SAFE ()
TODO


.SS BSD MACROS

For documentation of the BSD macros please refer to
.BR tree (3)


.SH EXAMPLE

.nf
#include <rb3ptr.h>

/*
 * Include the generic implementation. Alternatively, you can link with a
 * seperately compiled generic implementation
 */
RB3_GEN_IMPL_STATIC();

/*
 * Define a node datatype and a compare operation
 */
struct foo {
	/* the node type must include a struct rb3_head. */
	struct rb3_head head;
	int val;
}

int compare(struct foo *a, struct foo *b)
{
	return (a\->val > b\->val) \- (a\->val < b\->val);
}

RB3_GEN_INLINE_PROTO_STATIC(footree, struct foo, head);
RB3_GEN_NODECMP_PROTO_STATIC(footree, struct foo, head, compare);

void testoperations(void)
{
	struct footree tree;
	struct foo *iter;
	struct foo foo[42];
	size_t i;

	footree_init(&tree);
	for (i = 0; i < 42; i++)
		foo[i].val = rand();
	for (i = 0; i < 42; i++)
		footree_insert(&tree, &foo[i]);
	for (iter = footree_get_min(&tree); iter != NULL; iter = footree_get_next(iter))
		printf("iter %d\\n", iter\->val);
	for (i = 0; i < 42; i++)
		footree_delete(&tree, &foo[i]);
	footree_exit(&tree);
}
.fi


rb3ptr/rb3ptr.c
/*
 * Copyright (c) 2017 Jens Stimpfle
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <assert.h>
#include <stddef.h>  /* NULL */
#include "rb3ptr.h"

#define RB3_DIR_BIT (1<<0)
#define RB3_COLOR_BIT (1<<1)
#define RB3_BLACK (0)
#define RB3_RED (RB3_COLOR_BIT)
#define RB3_CHILD_PTR(head, color) ((rb3_ptr)(head) | color)
#define RB3_PARENT_PTR(head, dir) ((rb3_ptr)(head) | dir)

static struct rb3_head *rb3_get_black_child(struct rb3_head *head, int dir)
{
        return (struct rb3_head *) head->ptr[dir];
}

static int rb3_get_color_bit(struct rb3_head *head, int dir)
{
        return head->ptr[dir] & RB3_COLOR_BIT;
}

static int rb3_is_red(struct rb3_head *head, int dir)
{
        return rb3_get_color_bit(head, dir) != 0;
}

static void rb3_set_red(struct rb3_head *head, int dir)
{
        head->ptr[dir] |= RB3_COLOR_BIT;
}

static void rb3_set_black(struct rb3_head *head, int dir)
{
        head->ptr[dir] &= ~RB3_COLOR_BIT;
}

static void rb3_connect(struct rb3_head *head, int dir, struct rb3_head *child, int color)
{
        head->ptr[dir] = RB3_CHILD_PTR(child, color);
        child->ptr[RB3_PARENT] = RB3_PARENT_PTR(head, dir);
}

static void rb3_connect_null(struct rb3_head *head, int dir, struct rb3_head *child, int color)
{
        head->ptr[dir] = RB3_CHILD_PTR(child, color);
        if (child)
                child->ptr[RB3_PARENT] = RB3_PARENT_PTR(head, dir);
}

static void rb3_insert_rebalance(struct rb3_head *head)
{
        struct rb3_head *pnt;
        struct rb3_head *gpnt;
        struct rb3_head *ggpnt;
        int left;
        int right;
        int gdir;
        int ggdir;

        if (!rb3_get_parent(rb3_get_parent(head))) {
                rb3_set_black(rb3_get_parent(head), RB3_LEFT);
                return;
        }

        if (!rb3_is_red(rb3_get_parent(rb3_get_parent(head)), rb3_get_parent_dir(rb3_get_parent(head))))
                /* parent is black */
                return;

        /*
         * Since parent is red parent can't be the root.
         * So we have at least a grandparent node, and grand-grandparent
         * is either a real node or the base head.
         */
        pnt = rb3_get_parent(head);
        gpnt = rb3_get_parent(pnt);
        ggpnt = rb3_get_parent(gpnt);
        left = rb3_get_parent_dir(head);
        right = !rb3_get_parent_dir(head);
        gdir = rb3_get_parent_dir(pnt);
        ggdir = rb3_get_parent_dir(gpnt);

        if (rb3_is_red(gpnt, !gdir)) {
                /* uncle and parent are both red */
                rb3_set_red(ggpnt, ggdir);
                rb3_set_black(gpnt, RB3_LEFT);
                rb3_set_black(gpnt, RB3_RIGHT);
                rb3_insert_rebalance(gpnt);
        } else if (gdir == right) {
                rb3_connect_null(pnt, left, rb3_get_black_child(head, right), RB3_BLACK);
                rb3_connect_null(gpnt, right, rb3_get_black_child(head, left), RB3_BLACK);
                rb3_connect(head, left, gpnt, RB3_RED);
                rb3_connect(head, right, pnt, RB3_RED);
                rb3_connect(ggpnt, ggdir, head, RB3_BLACK);
        } else {
                rb3_connect_null(gpnt, left, rb3_get_black_child(pnt, right), RB3_BLACK);
                rb3_connect(pnt, right, gpnt, RB3_RED);
                rb3_connect(ggpnt, ggdir, pnt, RB3_BLACK);
        }
}

static void rb3_delete_rebalance(struct rb3_head *head)
{
        struct rb3_head *pnt;
        struct rb3_head *gpnt;
        struct rb3_head *sibling;
        struct rb3_head *sleft;
        struct rb3_head *sleftleft;
        struct rb3_head *sleftright;
        int left;
        int right;
        int gdir;

        if (!rb3_get_parent(head))
                return;

        if (!rb3_get_parent(rb3_get_parent(head)))
                return;

        pnt = rb3_get_parent(head);
        left = rb3_get_parent_dir(head);
        right = !rb3_get_parent_dir(head);
        gpnt = rb3_get_parent(pnt);
        gdir = rb3_get_parent_dir(pnt);
        sibling = rb3_get_child(pnt, right);
        sleft = rb3_get_child(sibling, left);

        if (rb3_is_red(pnt, right)) {
                /* sibling is red */
                rb3_connect(pnt, right, sleft, RB3_BLACK);
                rb3_connect(sibling, left, pnt, RB3_RED);
                rb3_connect(gpnt, gdir, sibling, RB3_BLACK);
                rb3_delete_rebalance(head);
        } else if (rb3_is_red(sibling, right)) {
                /* outer child of sibling is red */
                rb3_connect_null(pnt, right, sleft, rb3_get_color_bit(sibling, left));
                rb3_connect(sibling, left, pnt, RB3_BLACK);
                rb3_connect(gpnt, gdir, sibling, rb3_get_color_bit(gpnt, gdir));
                rb3_set_black(sibling, right);
        } else if (rb3_is_red(sibling, left)) {
                /* inner child of sibling is red */
                sleftleft = rb3_get_child(sleft, left);
                sleftright = rb3_get_child(sleft, right);
                rb3_connect_null(pnt, right, sleftleft, RB3_BLACK);
                rb3_connect_null(sibling, left, sleftright, RB3_BLACK);
                rb3_connect(sleft, left, pnt, RB3_BLACK);
                rb3_connect(sleft, right, sibling, RB3_BLACK);
                rb3_connect(gpnt, gdir, sleft, rb3_get_color_bit(gpnt, gdir));
        } else if (rb3_is_red(gpnt, gdir)) {
                /* parent is red */
                rb3_set_red(pnt, right);
                rb3_set_black(gpnt, gdir);
        } else {
                /* all relevant nodes are black */
                rb3_set_red(pnt, right);
                rb3_delete_rebalance(pnt);
        }
}

static void rb3_delete_noninternal(struct rb3_head *head)
{
        struct rb3_head *pnt;
        struct rb3_head *cld;
        int pdir;
        int dir;

        dir = rb3_get_child(head, RB3_RIGHT) ? RB3_RIGHT : RB3_LEFT;

        pnt = rb3_get_parent(head);
        cld = rb3_get_child(head, dir);
        pdir = rb3_get_parent_dir(head);

        if (!rb3_is_red(pnt, pdir) && !rb3_is_red(head, dir))
                /* To be deleted node is black (and child cannot be repainted)
                 * => height decreased */
                rb3_delete_rebalance(head);

        rb3_connect_null(pnt, pdir, cld, RB3_BLACK);
}

static void rb3_delete_internal(struct rb3_head *head)
{
        struct rb3_head *subst;
        struct rb3_head *left;
        struct rb3_head *right;
        struct rb3_head *parent;
        int pdir;
        int pcol;

        subst = rb3_get_descendant_successor(head);
        rb3_delete_noninternal(subst);
        *subst = *head;

        left = rb3_get_child(head, RB3_LEFT);
        right = rb3_get_child(head, RB3_RIGHT);
        parent = rb3_get_parent(head);
        pdir = rb3_get_parent_dir(head);
        pcol = rb3_get_color_bit(parent, pdir);

        if (left)
                left->ptr[RB3_PARENT] = RB3_PARENT_PTR(subst, RB3_LEFT);
        if (right)
                right->ptr[RB3_PARENT] = RB3_PARENT_PTR(subst, RB3_RIGHT);
        parent->ptr[pdir] = RB3_CHILD_PTR(subst, pcol);
}

void rb3_init(struct rb3_tree *tree)
{
        tree->base.ptr[RB3_LEFT] = RB3_CHILD_PTR(NULL, RB3_BLACK);
        tree->base.ptr[RB3_RIGHT] = RB3_CHILD_PTR(NULL, RB3_BLACK);
        tree->base.ptr[RB3_PARENT] = RB3_PARENT_PTR(NULL, RB3_BLACK);;
}

void rb3_exit(struct rb3_tree *tree)
{
        /* No resources allocated */
        (void) tree;
}

void rb3_insert(struct rb3_head *head, struct rb3_head *parent, int dir)
{
        assert(dir == RB3_LEFT || dir == RB3_RIGHT);

        parent->ptr[dir] = RB3_CHILD_PTR(head, RB3_RED);
        head->ptr[RB3_PARENT] = RB3_PARENT_PTR(parent, dir);
        head->ptr[RB3_LEFT] = RB3_CHILD_PTR(NULL, RB3_BLACK);
        head->ptr[RB3_RIGHT] = RB3_CHILD_PTR(NULL, RB3_BLACK);
        rb3_insert_rebalance(head);
}

void rb3_delete(struct rb3_head *head)
{
        if (rb3_has_child(head, RB3_LEFT) && rb3_has_child(head, RB3_RIGHT))
                rb3_delete_internal(head);
        else
                rb3_delete_noninternal(head);

        head->ptr[RB3_LEFT] = 0;
        head->ptr[RB3_RIGHT] = 0;
        head->ptr[RB3_PARENT] = 0;
}

struct rb3_head *rb3_get_min(struct rb3_tree *tree)
{
        struct rb3_head *head;

        head = rb3_get_root(tree);
        if (!head)
                return NULL;
        while (rb3_get_child(head, RB3_LEFT))
                head = rb3_get_child(head, RB3_LEFT);
        return head;
}

struct rb3_head *rb3_get_max(struct rb3_tree *tree)
{
        struct rb3_head *head;

        head = rb3_get_root(tree);
        if (!head)
                return NULL;
        while (rb3_get_child(head, RB3_RIGHT))
                head = rb3_get_child(head, RB3_RIGHT);
        return head;
}

struct rb3_head *rb3_get_ascendant_successor(struct rb3_head *head)
{
        while (head && rb3_get_parent_dir(head) == RB3_RIGHT)
                head = rb3_get_parent(head);
        if (head) {
                head = rb3_get_parent(head);
                if (head && !rb3_get_parent(head))
                        /* base fake element */
                        head = NULL;
        }
        return head;
}

struct rb3_head *rb3_get_ascendant_predecessor(struct rb3_head *head)
{
        while (head && rb3_get_parent_dir(head) == RB3_LEFT)
                head = rb3_get_parent(head);
        if (head)
                head = rb3_get_parent(head);
        return head;
}

struct rb3_head *rb3_get_descendant_successor(struct rb3_head *head)
{
        if (!rb3_has_child(head, RB3_RIGHT))
                return NULL;
        head = rb3_get_child(head, RB3_RIGHT);
        while (rb3_has_child(head, RB3_LEFT))
                head = rb3_get_child(head, RB3_LEFT);
        return head;
}

struct rb3_head *rb3_get_descendant_predecessor(struct rb3_head *head)
{
        if (!rb3_has_child(head, RB3_RIGHT))
                return NULL;
        head = rb3_get_child(head, RB3_RIGHT);
        while (rb3_get_child(head, RB3_LEFT))
                head = rb3_get_child(head, RB3_LEFT);
        return head;
}

struct rb3_head *rb3_get_inorder_successor(struct rb3_head *head)
{
        if (rb3_has_child(head, RB3_RIGHT))
                return rb3_get_descendant_successor(head);
        else
                return rb3_get_ascendant_successor(head);
}

struct rb3_head *rb3_get_inorder_predecessor(struct rb3_head *head)
{
        if (rb3_has_child(head, RB3_RIGHT))
                return rb3_get_descendant_predecessor(head);
        else
                return rb3_get_ascendant_predecessor(head);
}
